Support mbed_start_application for Cortex-M23

ccli8 opened the pull request:
### Description
This PR implements `mbed_start_application` on Cortex-M23 for bootloader related application.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] New target
    [ ] Feature
    [ ] Breaking change

ccli8 committed:
Support mbed_start_application for Cortex-M23

deepikabhavnani reviewed (dismissed):
Looks good to me

ccli8 committed:
Fix start_new_application compile error on Cortex-M23

With "mov r2, #0", compile OK with GCC_ARM, but failed with ARMC6.
With "ldr r2, =0", compile OK with ARMC6, but failed with GCC_ARM.
Finally, with "movw r2, #0"/"movt r2, #0", compile OK with both ARMC6 and GCC_ARM.

ccli8 commented:
In https://github.com/ARMmbed/mbed-os/pull/6949/commits/646f61400abb4d80aced65bbd286865fe68fea12, I fix `start_new_application` compile error with different toolchains:
1. With `mov r2, #0`, compile OK with **GCC_ARM**, but failed with **ARMC6**.
1. With `ldr r2, =0`, compile OK with **ARMC6**, but failed with **GCC_ARM**.
1. Finally, with `movw r2, #0`/`movt r2, #0`, compile OK with both **ARMC6** and **GCC_ARM**.

cmonr commented:
@dreemkiller Mind taking a look at this? Not sure who else would be familiar with the M23 aside for you and @deepikabhavnani

dreemkiller commented:
@TacoGrandeTX, take a look

TacoGrandeTX commented on platform/mbed_application.c:
```diff
@@ -48,12 +48,20 @@ static void powerdown_nvic()
     int j;
 
+#if defined(__CORTEX_M23)
+    isr_groups_32 = 16;
```
16 is not a good number... and more than even a M33 could support.  The max number of external interrupts a M23 can support is 240.  It's sad that M23 doesn't have the ICTR so the max value we could use here is 7 (or 8... depending on how it's used elsewhere).  Rule SGCR in v8-M ARM states:  "When a particular NVIC interrupt line is not implemented, the registers that are associated with it are reserved."  Yet Rule GLNG states "Privileged accesses to unimplemented registers are RES0" so accessing non-existing registers should be harmless but are wasteful.

ccli8 replied:
@TacoGrandeTX Thanks for your detailed information. I fix it in https://github.com/ARMmbed/mbed-os/pull/6949/commits/3c14cb633222cbc42fc9c1ea257da49db3a73d15.

TacoGrandeTX commented on platform/mbed_application.c:
```diff
@@ -69,17 +77,20 @@ static void powerdown_scb(uint32_t vtor)
     // SCB->CCR     - Implementation defined value
     for (i = 0; i < 12; i++) {
-#if defined(__CORTEX_M7)
+#if defined(__CORTEX_M7) || defined(__CORTEX_M23)
```
Not good - M23 has only 2 SHPR registers (the first is reserved).  Looks like we have mixed up the size of our register definitions... core_cm23.h is using uint32_t (x2) while core_cm7.h is using uint8_t (x12).  We need two for loops or should change SCB_Type in core_cm23.h.

ccli8 replied:
@TacoGrandeTX I fix it in https://github.com/ARMmbed/mbed-os/pull/6949/commits/3c14cb633222cbc42fc9c1ea257da49db3a73d15.

TacoGrandeTX commented on platform/mbed_application.c:
```diff
@@ -107,7 +118,12 @@ __asm static void start_new_application(void *sp, void *pc)
 void start_new_application(void *sp, void *pc)
 {
     __asm volatile (
+#if defined(__CORTEX_M23)
```
This two-instruction sequence is not required... "mov r2, #0" will put 0 (zero-extended) into r2 on the M23.  We can remove the entire #if block.  MOVW/MOVT is useful for loading a full 32-value into a register without a literal pool... not the case here.

TacoGrandeTX replied:
Sorry - missed the commit comments!  Clearly AC6 is not accepting the inline assembly properly (I ran a local test) so perhaps we could use "movw r2, #0" for both conditions and still remove the #if block.

ccli8 replied:
@TacoGrandeTX I replace with `movw`/`movt` and remove `#if` block in https://github.com/ARMmbed/mbed-os/pull/6949/commits/3c14cb633222cbc42fc9c1ea257da49db3a73d15.

cmonr commented:
@ccli8 ^^^ ?

cmonr commented:
Taking another look at this PR, this can come in with a patch release.

ccli8 committed:
Fix mbed_start_application on Cortex-M23

1. M23 doesn't support ICTR and supports up to 240 external interrupts.
2. Fix reset of SHPR
3. Fix inline assembly compile error with ARMC6

cmonr commented:
@TacoGrandeTX @deepikabhavnani Mind re-reviewing?

TacoGrandeTX commented on platform/mbed_application.c:
```diff
@@ -118,12 +125,8 @@ __asm static void start_new_application(void *sp, void *pc)
-#if defined(__CORTEX_M23)
-        "movw   r2, #0      \n"
+        "movw   r2, #0      \n" // Fail to compile "mov r2, #0" with ARMC6. Replace with movw/movt.
         "movt   r2, #0      \n"
```
Sorry - my previous comments weren't very clear!  MOVT is superfluous here as MOVW will zero-extend.  So we can remove the MOVT instruction.  "MOVW r2, #0" will put 0x0 into all 32-bits of r2.

ccli8 replied:
@TacoGrandeTX Yes, **MOVT** is superfluous. I remove it in https://github.com/ARMmbed/mbed-os/pull/6949/commits/23dcd82fabcb81a40078fc904d604d63997aca6c.

TacoGrandeTX reviewed (commented):
MOVT is superfluous - recommend removing it.

ccli8 committed:
Remove superfluous MOVT in mbed_start_application/start_new_application

cmonr commented:
Looks like all of the review comments have been addressed. 
@TacoGrandeTX Mind confirming?

/morph build

adbridge commented:
Restarted pr-head

cmonr commented:
Interesting... Don't think the test failure is the PR's fault.
/morph test

cmonr commented:
@TacoGrandeTX Could you re-approve if you're alright with the changes?

TacoGrandeTX reviewed (approved):
Good!

cmonr commented:
/morph build

cmonr commented:
Oh...
Me and my reflexes.
